src/main/java/org/truffleruby/core/string/TStringCache.java

```diff
@@ -69,20 +70,37 @@ private void register(TruffleString tstring, RubyEncoding encoding) {
         }
     }
 
-    public TruffleString getTString(TruffleString string, RubyEncoding encoding) {
-        return getTString(TStringUtils.getBytesOrCopy(string, encoding), encoding);
+    @TruffleBoundary
+    public TruffleString getTString(TruffleString string, RubyEncoding rubyEncoding) {
+        assert rubyEncoding != null;
+
+        var byteArray = string.getInternalByteArrayUncached(rubyEncoding.tencoding);
+        final TBytesKey key = new TBytesKey(byteArray, TStringUtils.hasImmutableInternalByteArray(string),
+                rubyEncoding);
+
+        return getTString(key);
+    }
+
+    public TruffleString getTString(InternalByteArray byteArray, boolean isImmutable, RubyEncoding rubyEncoding) {
+        assert rubyEncoding != null;
+
+        return getTString(new TBytesKey(byteArray, isImmutable, rubyEncoding));
     }
 
-    @TruffleBoundary
     public TruffleString getTString(byte[] bytes, RubyEncoding rubyEncoding) {
         assert rubyEncoding != null;
 
-        final TBytesKey key = new TBytesKey(bytes, rubyEncoding);
+        return getTString(new TBytesKey(bytes, rubyEncoding));
+    }
 
+    @TruffleBoundary
+    private TruffleString getTString(TBytesKey key) {
```

Comment:
Could you name this one `lookupKey` to make it clear it should only be used for lookup, not for adding in the map?